File: dichvucong_extension/test_csv.js
const fs = require('fs');

function parseAndFilterCSV(text) {
    const lines = text.split('\n');
    if (lines.length < 2) return [];

    // Tìm index của các cột bằng Regex (bảo vệ dấu phẩy nằm trong ngoặc kép)
    const headers = lines[0].split(/,(?=(?:(?:[^"]*"){2})*[^"]*$)/).map(h => h.trim().replace(/^"|"$/g, ''));
    console.log("Headers length:", headers.length);
    const nameIdx = headers.findIndex(h => h === 'Họ và tên');
    const idIdx = headers.findIndex(h => h.includes('Số CC') || h.includes('định danh') || h === 'Số CC/ CCCD/ Số định danh');
    const checkinIdx = headers.findIndex(h => h === 'Check in');
    const statusIdx = headers.findIndex(h => h === 'Trang thai' || h === 'Trạng thái');

    console.log("Found Indexes:", { nameIdx, idIdx, checkinIdx, statusIdx });

    if (nameIdx === -1 || idIdx === -1) {
        console.log('Error: Missing Name or ID column');
        return [];
    }

    const results = [];
    for (let i = 1; i < Math.min(lines.length, 5); i++) {
        if (!lines[i].trim()) continue;

        const row = lines[i].split(/,(?=(?:(?:[^"]*"){2})*[^"]*$)/).map(item => item.trim().replace(/^"|"$/g, ''));

        const checkin = (row[checkinIdx] || '').trim();
        const status = (row[statusIdx] || '').trim();

        console.log(`Row ${i}: Checkin='${checkin}', Status='${status}'`);

        if (checkin.toUpperCase() === 'Y' && status.toLowerCase() === 'dat') {
            results.push({
                hoTen: row[nameIdx],
                soCCCD: row[idIdx],
                lyDo: 'Tu Học'
            });
        }
    }
    return results;
}

const data = fs.readFileSync('/Users/sonpc/Downloads/Thông tin đăng ký khoá tu tại chùa Từ Đức - Data.csv', 'utf8');
const res = parseAndFilterCSV(data);
console.log("Final matched results:", res.length);

File: dichvucong_extension/test_csv.py
import csv

def parse_csv(file_path):
    print(f"Reading file: {file_path}")
    results = []
    
    with open(file_path, newline='', encoding='utf-8') as csvfile:
        reader = csv.reader(csvfile)
        headers = next(reader)
        
        headers = [h.strip() for h in headers]
        
        try:
            name_idx = headers.index('Họ và tên')
            
            # Find ID index
            id_idx = -1
            for i, h in enumerate(headers):
                if 'Số CC' in h or 'định danh' in h.lower():
                    id_idx = i
                    break
                    
            checkin_idx = headers.index('Check in')
            status_idx = headers.index('Trạng thái') if 'Trạng thái' in headers else headers.index('Trang thai')
            
            print(f"Indexes: Name={name_idx}, ID={id_idx}, Checkin={checkin_idx}, Status={status_idx}")
            
        except ValueError as e:
            print("Lỗi nhận diện cột:", e)
            print("Headers:", headers)
            return

        for i, row in enumerate(reader):
            if i >= 5: break
            
            if len(row) > max(name_idx, id_idx, checkin_idx, status_idx):
                checkin = row[checkin_idx].strip()
                status = row[status_idx].strip()
                print(f"Row {i+1}: Name='{row[name_idx]}', Checkin='{checkin}', Status='{status}'")
                
                if checkin.upper() == 'Y' and status.lower() == 'dat':
                    results.append(row[name_idx])
                    
    print(f"\nFinal Matched Rows: {len(results)}")

parse_csv('/Users/sonpc/Downloads/Thông tin đăng ký khoá tu tại chùa Từ Đức - Data.csv')
